modules/gateways/Blockonomics/Blockonomics.php

```diff
@@ -318,63 +372,243 @@ public function insertOrderToDb($id_order, $address, $value, $bits) {
+			if($order->status == 0 || $order->status == 1){
+				return $order;
+			}
+		}
+	}
+
+	/*
+	 * Check for unused address for the currency linked to the order
+	 */		
+	public function getAndUpdateWaitingOrder($orders, $blockonomics_currency) {
+		foreach ($orders as $order) {
+			//check for currency address already waiting
+			if($order->blockonomics_currency == $blockonomics_currency && $order->status == -1){
+				//Check if existing order is expired
+				$current_time = time();
+				$total_time = $this->getTimePeriod() * 60;
+				$clock = $order->timestamp + $total_time - $current_time;
+				if($clock < 0){
+					$order->bits = $this->convertFiatToBlockonomicsCurrency($order->value, $order->blockonomics_currency);
+					$order->timestamp = $current_time;
+				}
+				$this->updateOrderExpected($order->addr, $order->blockonomics_currency, $order->timestamp, $order->bits);
+				return $order;
+			}
+		}
+	}
+
+	/*
+	 * Check for unused address or create new
+	 */	
+	public function processNewOrder($orders, $blockonomics_currency) {
+		foreach ($orders as $order) {
+			//check for new blank order in db
+			if($order->addr == ""){
+				$new_addresss_response = $this->getNewAddress($blockonomics_currency);
+				if ($new_addresss_response->response_code == 200){
+					$order->addr = $new_addresss_response->address;
+				}else{
+					exit($new_addresss_response->message);
+				}
+
+				$this->updateOrderAddress($order->id_order, $order->addr, $blockonomics_currency);
+				$order->blockonomics_currency = $blockonomics_currency;
+				$order->bits = $this->convertFiatToBlockonomicsCurrency($order->value, $order->blockonomics_currency);
+				$order->timestamp = time();
+				$this->updateOrderExpected($order->addr, $order->blockonomics_currency, $order->timestamp, $order->bits);
+				return $order;
+			}
+		}
+		// blank order already used, create a new blank order with same uuid
+		// fix this
+		$order_uuid = $this->newOrder($orders[0]->value, $orders[0]->id_order);
+		$order = $this->processOrderHash($order_uuid, $blockonomics_currency);
+		return $order;
+	}
+
+	/*
+	 * Find an existing order or create a new order
+	 */	
+	public function processOrderHash($order_hash, $blockonomics_currency) {
+		// Fetch all orders by hash
+		$all_orders_by_hash = $this->getAllOrdersByHash($order_hash);
+		if(!$all_orders_by_hash){
+			exit;
+		}
+		// Check for pending payments and return the order
+		$pending_payment = $this->getPendingOrder($all_orders_by_hash);
+		if($pending_payment){
+			return $pending_payment;
+		}
+		// Check for existing address
+		$address_waiting = $this->getAndUpdateWaitingOrder($all_orders_by_hash, $blockonomics_currency);
+		if($address_waiting){
+			$address_waiting->currency = getCurrency(getClientsDetails()['user_id'])['code'];
+			return $address_waiting;
+		}
+		// Process a new order for the hash and blockonomics currency
+		$new_order = $this->processNewOrder($all_orders_by_hash, $blockonomics_currency);
```

Comment:
Comments on processNewOrder. The whole process of creating order with empty address is unnecessary  as it creates empty rows. Let us do the following:
-  Rename *processNewOrder* to *createNewCryptoOrder* and pass it order_id, order_value and blockonomics_currency. No need to loop just take these values from $all_orders_by_hash[0]
-  *updateOrderAddress* will first search order having empty  addr. If found it will use that, else it will create new row with iinput order_value, order_id and address and blockonomics_currency